lib/recursive-methods.rb

```diff
@@ -1,49 +1,121 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), where n is the size of the input number
+# Space complexity: O(n),  where n is the size of the input number
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+
+    if n < 0
+        raise ArgumentError, "number can't be negative"
+    end
+    if n == 0 
+        return 1
+    end
+    return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), where n is the length of the string
```

Comment:
This is not in place since you create a new string. 

Think about making a helper which takes the string, a left index and right index.  The base case is when `left_index >= right_index`  Each recursive call you swap the characters at `left_index` or `right_index` and then make a recursive call with the string, and `left_index + 1` and `right_index -1`.